### PR TITLE
Support rejecting invalid requests

### PR DESCRIPTION
**Pull request by rafax**
This PR adds support for rejecting invalid requests with a 400 Bad Request error. 
For context, see this [discussion](https://sourcegraph.slack.com/archives/C0C324C91/p1696322332990229) and this [doc](https://docs.google.com/document/d/1KUnjMmQnQaZXQ3PFpd6mrKzGAE6xUfh9eKy4G2yhXyE/edit#heading=h.d2zpmd9eatt3).
## Test plan

- unit tests
- local testing
- blocking disabled by default

### CODE REVIEW

File: cmd/cody-gateway/shared/config.go [hunk 1]
```diff
@@ -113,6 +114,7 @@ func (c *Config) Load() {
 	}
 	c.Anthropic.MaxTokensToSample = c.GetInt("CODY_GATEWAY_ANTHROPIC_MAX_TOKENS_TO_SAMPLE", "10000", "Maximum permitted value of maxTokensToSample")
 	c.Anthropic.AllowedPromptPatterns = splitMaybe(c.GetOptional("CODY_GATEWAY_ANTHROPIC_ALLOWED_PROMPT_PATTERNS", "Prompt patterns to allow."))
+	c.Anthropic.RequestBlockingEnabled = c.GetBool("CODY_GATEWAY_ANTHROPIC_REQUEST_BLOCKING_ENABLED", "false", "Whether we should block requests.")
```
Comment (willdollman) [inline cmd/cody-gateway/shared/config.go:117]:
```suggestion
	c.Anthropic.RequestBlockingEnabled = c.GetBool("CODY_GATEWAY_ANTHROPIC_REQUEST_BLOCKING_ENABLED", "false", "Whether we should block requests that match our blocking criteria.")
```

**Review by willdollman — APPROVED** [review]

File: cmd/cody-gateway/internal/httpapi/completions/anthropic.go [hunk 2]
```diff
@@ -27,8 +27,11 @@ const anthropicAPIURL = "https://api.anthropic.com/v1/complete"
 const (
 	logPromptPrefixLength = 250
 
-	promptTokenLimit   = 18000
-	responseTokenLimit = 1000
+	promptTokenFlaggingLimit   = 18000
+	responseTokenFlaggingLimit = 1000
+
+	promptTokenBlockingLimit   = 20000
+	responseTokenBlockingLimit = 1000
```
[the remark this answers is not in the document]
Comment (rafax) [reply cmd/cody-gateway/internal/httpapi/completions/anthropic.go:34]: Fair point - I tried to follow the existing approach, and Cloud Run deployments are cheap and easy (and config changes require a deployment too) so I assumed this is ok. I will keep this in mind if we end up tweaking those too often without changing code.

[the remark this answers is not in the document]
Comment (rafax) [reply cmd/cody-gateway/internal/httpapi/completions/anthropic.go:34]: Also, I have a question about Cody Gateway config in general, will ask on Slack.

File: cmd/cody-gateway/internal/httpapi/completions/anthropic.go [hunk 3]
```diff
@@ -132,6 +142,10 @@ func NewAnthropicHandler(
 					if err := promptRecorder.Record(ctx, ar.Prompt); err != nil {
 						logger.Warn("failed to record flagged prompt", log.Error(err))
 					}
+					if requestBlockingEnabled && result.shouldBeBlocked {
+						result.wasBlocked = true
+						return http.StatusBadRequest, result, errors.Errorf("request blocked")
```
[the remark this answers is not in the document]
Comment (rafax) [reply cmd/cody-gateway/internal/httpapi/completions/anthropic.go]: Done

**Comment by rafax** [comment]
@willdollman @eseliger I re-requested your review since I added a new event to track blocked requests.

**Review by eseliger — APPROVED** [review]

**Review by willdollman — APPROVED** [review]

**Comment by bobheadxi** [comment]
Sorry I missed this PR, thanks for building this!

**Pull request merged**